File: API/1886_Sma_Trend_Filter/CS/SmaTrendFilterStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Multi-timeframe SMA trend filter strategy.
/// </summary>
public class SmaTrendFilterStrategy : Strategy
{
	private readonly StrategyParam<int> _openLevel;
	private readonly StrategyParam<int> _closeLevel;
	private readonly StrategyParam<DataType> _candleType1;
	private readonly StrategyParam<DataType> _candleType2;
	private readonly StrategyParam<DataType> _candleType3;
	
	private readonly int[] _periods = { 5, 8, 13, 21, 34 };
	private readonly Sma[][] _smas = new Sma[3][];
	private readonly decimal[][] _previous = new decimal[3][];
	private readonly decimal[] _uitog = new decimal[3];
	private readonly decimal[] _ditog = new decimal[3];
	private int _signal;
	
	/// <summary>
	/// Signal threshold to open position.
	/// </summary>
	public int OpenLevel
	{
		get => _openLevel.Value;
		set => _openLevel.Value = value;
}

/// <summary>
/// Signal threshold to close position.
/// </summary>
public int CloseLevel
{
	get => _closeLevel.Value;
	set => _closeLevel.Value = value;
}

/// <summary>
/// Primary timeframe for calculations.
/// </summary>
public DataType CandleType1
{
	get => _candleType1.Value;
	set => _candleType1.Value = value;
}

/// <summary>
/// Secondary timeframe for calculations.
/// </summary>
public DataType CandleType2
{
	get => _candleType2.Value;
	set => _candleType2.Value = value;
}

/// <summary>
/// Tertiary timeframe for calculations.
/// </summary>
public DataType CandleType3
{
	get => _candleType3.Value;
	set => _candleType3.Value = value;
}

/// <summary>
/// Initializes a new instance of the strategy.
/// </summary>
public SmaTrendFilterStrategy()
{
	_openLevel = Param(nameof(OpenLevel), 0)
	.SetDisplay("Open Level", "Signal threshold to open position", "Trading");
	_closeLevel = Param(nameof(CloseLevel), 0)
	.SetDisplay("Close Level", "Signal threshold to close position", "Trading");
	_candleType1 = Param(nameof(CandleType1), TimeSpan.FromMinutes(15).TimeFrame())
	.SetDisplay("Candle Type 1", "Primary timeframe", "General");
	_candleType2 = Param(nameof(CandleType2), TimeSpan.FromHours(1).TimeFrame())
	.SetDisplay("Candle Type 2", "Secondary timeframe", "General");
	_candleType3 = Param(nameof(CandleType3), TimeSpan.FromHours(4).TimeFrame())
	.SetDisplay("Candle Type 3", "Tertiary timeframe", "General");
	
	for (var i = 0; i < 3; i++)
	{
		_smas[i] = new Sma[_periods.Length];
		_previous[i] = new decimal[_periods.Length];
		for (var j = 0; j < _periods.Length; j++)
		_smas[i][j] = new Sma { Length = _periods[j] };
}
}

/// <inheritdoc />
public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
{
	return [(Security, CandleType1), (Security, CandleType2), (Security, CandleType3)];
}

/// <inheritdoc />
protected override void OnStarted(DateTimeOffset time)
{
	base.OnStarted(time);
	
	var sub1 = SubscribeCandles(CandleType1);
	sub1.Bind(_smas[0][0], _smas[0][1], _smas[0][2], _smas[0][3], _smas[0][4], ProcessTf1).Start();
	
	var sub2 = SubscribeCandles(CandleType2);
	sub2.Bind(_smas[1][0], _smas[1][1], _smas[1][2], _smas[1][3], _smas[1][4], ProcessTf2).Start();
	
	var sub3 = SubscribeCandles(CandleType3);
	sub3.Bind(_smas[2][0], _smas[2][1], _smas[2][2], _smas[2][3], _smas[2][4], ProcessTf3).Start();
}

private void ProcessTf1(ICandleMessage candle, decimal sma5, decimal sma8, decimal sma13, decimal sma21, decimal sma34)
=> ProcessTf(0, candle, new[] { sma5, sma8, sma13, sma21, sma34 });

private void ProcessTf2(ICandleMessage candle, decimal sma5, decimal sma8, decimal sma13, decimal sma21, decimal sma34)
=> ProcessTf(1, candle, new[] { sma5, sma8, sma13, sma21, sma34 });

private void ProcessTf3(ICandleMessage candle, decimal sma5, decimal sma8, decimal sma13, decimal sma21, decimal sma34)
=> ProcessTf(2, candle, new[] { sma5, sma8, sma13, sma21, sma34 });

private void ProcessTf(int index, ICandleMessage candle, decimal[] values)
{
	if (candle.State != CandleStates.Finished)
	return;
	
	var up = 0;
	var down = 0;
	
	for (var i = 0; i < _periods.Length; i++)
	{
		var val = values[i];
		
		if (val == 0m)
		return;
		
		var prev = _previous[index][i];
		
		if (prev == 0m)
		{
			_previous[index][i] = val;
			return;
	}
	
	if (val > prev)
	up++;
	else if (val < prev)
	down++;
	
	_previous[index][i] = val;
}

_uitog[index] = up / (decimal)_periods.Length * 100m;
_ditog[index] = down / (decimal)_periods.Length * 100m;

EvaluateSignal();
}

private void EvaluateSignal()
{
	_signal = 0;
	
	if (_uitog[0] >= 75m && _uitog[1] >= 75m && _uitog[2] >= 75m)
	_signal = 2;
	else if (_ditog[0] >= 75m && _ditog[1] >= 75m && _ditog[2] >= 75m)
	_signal = -2;
	else if (_uitog[0] > 50m && _uitog[1] > 50m && _uitog[2] > 50m)
	_signal = 1;
	else if (_ditog[0] > 50m && _ditog[1] > 50m && _ditog[2] > 50m)
	_signal = -1;
	
	if (!IsFormedAndOnlineAndAllowTrading())
	return;
	
	var openBuy = _signal > OpenLevel;
	var openSell = _signal < -OpenLevel;
	var closeBuy = _signal < -CloseLevel;
	var closeSell = _signal > CloseLevel;
	
	if (Position > 0 && closeBuy)
	SellMarket();
	
	if (Position < 0 && closeSell)
	BuyMarket();
	
	if (openBuy && Position <= 0 && !closeBuy)
	BuyMarket();
	
	if (openSell && Position >= 0 && !closeSell)
	SellMarket();
}
}

File: API/1886_Sma_Trend_Filter/PY/sma_trend_filter_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage
from StockSharp.Algo.Strategies import Strategy


class sma_trend_filter_strategy(Strategy):
    def __init__(self):
        super(sma_trend_filter_strategy, self).__init__()
        self._cooldown_bars = self.Param("CooldownBars", 200) \
            .SetDisplay("Cooldown Bars", "Minimum number of primary timeframe bars between orders", "Trading")
        self._candle_type1 = self.Param("CandleType1", DataType.TimeFrame(TimeSpan.FromMinutes(5))) \
            .SetDisplay("Candle Type 1", "Primary timeframe", "General")
        self._candle_type2 = self.Param("CandleType2", DataType.TimeFrame(TimeSpan.FromMinutes(15))) \
            .SetDisplay("Candle Type 2", "Secondary timeframe", "General")
        self._candle_type3 = self.Param("CandleType3", DataType.TimeFrame(TimeSpan.FromHours(1))) \
            .SetDisplay("Candle Type 3", "Tertiary timeframe", "General")

        self._periods = [5, 8, 13, 21, 34]
        self._smas = [[None] * 5 for _ in range(3)]
        # Previous values: [tf_index][sma_index]
        self._previous = [[0.0] * 5 for _ in range(3)]
        self._uitog = [0.0] * 3
        self._ditog = [0.0] * 3
        self._is_ready = [False] * 3
        self._signal = 0
        self._bars_since_trade = 0

        for i in range(3):
            for j in range(5):
                sma = SimpleMovingAverage()
                sma.Length = self._periods[j]
                self._smas[i][j] = sma

    @property
    def cooldown_bars(self):
        return self._cooldown_bars.Value

    @property
    def candle_type1(self):
        return self._candle_type1.Value

    @property
    def candle_type2(self):
        return self._candle_type2.Value

    @property
    def candle_type3(self):
        return self._candle_type3.Value

    def OnReseted(self):
        super(sma_trend_filter_strategy, self).OnReseted()
        self._signal = 0
        self._bars_since_trade = 0
        for i in range(3):
            self._uitog[i] = 0.0
            self._ditog[i] = 0.0
            self._is_ready[i] = False
            for j in range(5):
                self._previous[i][j] = 0.0
                self._smas[i][j].Reset()

    def OnStarted2(self, time):
        super(sma_trend_filter_strategy, self).OnStarted2(time)
        sub1 = self.SubscribeCandles(self.candle_type1)
        sub1.Bind(self._smas[0][0], self._smas[0][1], self._smas[0][2], self._smas[0][3], self._smas[0][4], self.process_tf1).Start()
        sub2 = self.SubscribeCandles(self.candle_type2)
        sub2.Bind(self._smas[1][0], self._smas[1][1], self._smas[1][2], self._smas[1][3], self._smas[1][4], self.process_tf2).Start()
        sub3 = self.SubscribeCandles(self.candle_type3)
        sub3.Bind(self._smas[2][0], self._smas[2][1], self._smas[2][2], self._smas[2][3], self._smas[2][4], self.process_tf3).Start()

    def process_tf1(self, candle, v0, v1, v2, v3, v4):
        if candle.State != CandleStates.Finished:
            return
        self._process_tf_values(0, v0, v1, v2, v3, v4)
        if self._is_ready[0]:
            self._bars_since_trade += 1
            self._evaluate_signal()

    def process_tf2(self, candle, v0, v1, v2, v3, v4):
        if candle.State != CandleStates.Finished:
            return
        self._process_tf_values(1, v0, v1, v2, v3, v4)

    def process_tf3(self, candle, v0, v1, v2, v3, v4):
        if candle.State != CandleStates.Finished:
            return
        self._process_tf_values(2, v0, v1, v2, v3, v4)

    def _process_tf_values(self, tf_index, v0, v1, v2, v3, v4):
        vals = [float(v0), float(v1), float(v2), float(v3), float(v4)]

        for i in range(5):
            if vals[i] == 0.0:
                return

        # Get previous values
        prevs = [self._previous[tf_index][i] for i in range(5)]

        is_ready = True
        for i in range(5):
            if prevs[i] == 0.0:
                is_ready = False

        # Store current values
        for i in range(5):
            self._previous[tf_index][i] = vals[i]

        if not is_ready:
            return

        up = 0
        down = 0
        for i in range(5):
            if vals[i] > prevs[i]:
                up += 1
            elif vals[i] < prevs[i]:
                down += 1

        up_pct = up / 5.0 * 100.0
        down_pct = down / 5.0 * 100.0

        self._uitog[tf_index] = up_pct
        self._ditog[tf_index] = down_pct
        self._is_ready[tf_index] = True

    def _evaluate_signal(self):
        if not self._is_ready[0] or not self._is_ready[1] or not self._is_ready[2]:
            return

        # Compute average trend strength across all 3 timeframes
        avg_up = (self._uitog[0] + self._uitog[1] + self._uitog[2]) / 3.0
        avg_down = (self._ditog[0] + self._ditog[1] + self._ditog[2]) / 3.0

        # Determine signal based on average trend strength
        self._signal = 0

        if avg_up >= 80.0:
            self._signal = 2
        elif avg_down >= 80.0:
            self._signal = -2
        elif avg_up >= 60.0:
            self._signal = 1
        elif avg_down >= 60.0:
            self._signal = -1

        if self._bars_since_trade < self.cooldown_bars:
            return

        # Close logic: close when signal reverses strongly
        if self.Position > 0 and self._signal <= -2:
            self.SellMarket()
            self._bars_since_trade = 0
            return

        if self.Position < 0 and self._signal >= 2:
            self.BuyMarket()
            self._bars_since_trade = 0
            return

        # Open logic: open when signal is strong (level 2)
        if self._signal >= 2 and self.Position <= 0:
            self.BuyMarket()
            self._bars_since_trade = 0
        elif self._signal <= -2 and self.Position >= 0:
            self.SellMarket()
            self._bars_since_trade = 0

    def CreateClone(self):
        return sma_trend_filter_strategy()
